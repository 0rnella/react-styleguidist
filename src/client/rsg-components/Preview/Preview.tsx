import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import Context from 'rsg-components/Context';
import ReactExample from 'rsg-components/ReactExample';
import PlaygroundError from 'rsg-components/PlaygroundError';

interface PreviewProps {
	code: string;
	filepath: string;
	index: number;
	evalInContext(code: string): () => any;
}

interface PreviewState {
	error: string | null;
	height: number;
}

export default class Preview extends Component<PreviewProps, PreviewState> {
	public static propTypes = {
		code: PropTypes.string.isRequired,
		filepath: PropTypes.string.isRequired,
		evalInContext: PropTypes.func.isRequired,
	};
	public static contextType = Context;

	public state: PreviewState = {
		error: null,
		height: 20,
	};

	private iframeRef = React.createRef<HTMLIFrameElement>();

	public componentDidMount() {
		// Clear console after hot reload, do not clear on the first load
		// to keep any warnings
		if (this.context.codeRevision > 0) {
			// eslint-disable-next-line no-console
			console.clear();
		}

		this.executeCode();
	}

	public shouldComponentUpdate(nextProps: PreviewProps, nextState: PreviewState) {
		return (
			this.state.error !== nextState.error ||
			this.state.height !== nextState.height ||
			this.props.code !== nextProps.code
		);
	}

	public componentDidUpdate(prevProps: PreviewProps) {
		if (this.props.code !== prevProps.code) {
			this.executeCode();
		}
	}

	public componentWillUnmount() {
		this.unmountPreview();
	}

	private getMountNode() {
		// TODO: Do not render to body
		return this.iframeRef.current?.contentWindow?.document?.body;
	}

	private unmountPreview() {
		const node = this.getMountNode();
		if (!node) {
			return;
		}
		ReactDOM.unmountComponentAtNode(node);
	}

	private executeCode() {
		this.setState({
			error: null,
		});

		const { code } = this.props;
		if (!code) {
			return;
		}

		const wrappedComponent: React.FunctionComponentElement<any> = (
			<ReactExample
				code={code}
				evalInContext={this.props.evalInContext}
				onError={this.handleError}
				compilerConfig={this.context.config.compilerConfig}
			/>
		);

		window.requestAnimationFrame(() => {
			const node = this.getMountNode();
			if (!node) {
				return;
			}

			// this.unmountPreview();
			try {
				ReactDOM.render(wrappedComponent, node);
				this.createMutationObserver(node);
				this.handleResize();
			} catch (err) {
				this.handleError(err);
			}
		});
	}

	private createMutationObserver(node: HTMLElement) {
		// TODO: Destroy after unmount
		const observer = new MutationObserver(this.handleResize);
		observer.observe(node, {
			attributes: true,
			attributeOldValue: false,
			characterData: true,
			characterDataOldValue: false,
			childList: true,
			subtree: true,
		});
	}

	// TOOD: Debounce
	private handleResize = () => {
		const node = this.getMountNode();
		if (!node) {
			return;
		}

		this.setState(prevState => {
			const height = node.scrollHeight;
			if (prevState.height === height) {
				return null;
			}
			return { height };
		});
	};

	private handleError = (err: Error) => {
		this.unmountPreview();

		this.setState({
			error: err.toString(),
			height: 0,
		});

		console.error(err); // eslint-disable-line no-console
	};

	public render() {
		const { height, error } = this.state;
		// TODO: Global styles
		// TODO: title
		return (
			<>
				<iframe
					data-testid="mountNode"
					title="TODO"
					ref={this.iframeRef}
					height={height}
					frameBorder={0}
				></iframe>
				{error && <PlaygroundError message={error} />}
			</>
		);
	}
}
