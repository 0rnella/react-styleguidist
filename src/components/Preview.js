// Based on https://github.com/joelburget/react-live-editor/blob/master/live-compile.jsx

import React, { PropTypes } from 'react';
import reactTools from 'react-tools';

export default React.createClass({
	displayName: 'Preview',
	propTypes: {
		code: PropTypes.string.isRequired
	},

	getInitialState() {
		return {
			error: null
		};
	},

	componentDidMount: function() {
		this.executeCode();
	},

	componentDidUpdate: function(prevProps) {
		if (this.props.code !== prevProps.code) {
			this.executeCode();
		}
	},

	compileCode: function() {
		// TODO: Babel
		return reactTools.transform(
			// '(function() {' +
				this.props.code,
			// '\n})();'
			{
				harmony: true
			}
		);
	},

	executeCode: function() {
		var mountNode = this.refs.mount.getDOMNode();

		try {
			React.unmountComponentAtNode(mountNode);
		}
		catch (e) {
		}

		this.setState({
			error: null
		});

		try {
			var compiledCode = this.compileCode();
			React.render(eval(compiledCode), mountNode);
		}
		catch (err) {
			React.unmountComponentAtNode(mountNode);
			this.setState({
				error: err.toString()
			});
		}
	},

	renderError() {
		let { error } = this.state;
		if (error) {
			return (
				<div className="playgroundError">{error}</div>
			);
		}
		else {
			return null;
		}
	},

	render: function() {
		return (
			<div>
				<div ref="mount"></div>
				{this.renderError()}
			</div>
		);
	}
});
