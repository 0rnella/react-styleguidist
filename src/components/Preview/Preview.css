.playgroundError {
	margin: -1rem -1rem -1.2rem;
	padding: 1rem;
	line-height: 1.2;
	font-size: .9rem;
	color: #fff;
	background: #c00;
}
