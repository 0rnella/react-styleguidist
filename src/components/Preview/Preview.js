// Based on https://github.com/joelburget/react-live-editor/blob/master/live-compile.jsx

import React, { Component, PropTypes } from 'react';
import ReactDOM from 'react-dom';
import reactTools from 'react-tools';
import Wrapper from 'components/Wrapper';

import s from './Preview.css';

export default class Preview extends Component {
	static propTypes = {
		code: PropTypes.string.isRequired
	}

	constructor() {
		super();
		this.state = {
			error: null
		};
	}

	componentDidMount() {
		this.executeCode();
	}

	componentDidUpdate(prevProps) {
		if (this.props.code !== prevProps.code) {
			this.executeCode();
		}
	}

	compileCode() {
		// TODO: Babel
		return reactTools.transform(this.props.code, {
			harmony: true
		});
	}

	executeCode() {
		let mountNode = this.refs.mount;

		try {
			ReactDOM.unmountComponentAtNode(mountNode);
		}
		finally {
			/* */
		}

		this.setState({
			error: null
		});

		try {
			let compiledCode = this.compileCode();
			let component = eval(compiledCode);  /* eslint no-eval:0 */
			let wrappedComponent = (
				<Wrapper>
					{component}
				</Wrapper>
			);
			ReactDOM.render(wrappedComponent, mountNode);
		}
		catch (err) {
			ReactDOM.unmountComponentAtNode(mountNode);
			this.setState({
				error: err.toString()
			});
		}
	}

	renderError() {
		let { error } = this.state;
		if (error) {
			return (
				<div className={s.playgroundError}>{error}</div>
			);
		}
		else {
			return null;
		}
	}

	render() {
		return (
			<div>
				<div ref="mount"></div>
				{this.renderError()}
			</div>
		);
	}
}
