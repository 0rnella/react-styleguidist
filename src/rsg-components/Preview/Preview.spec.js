import React from 'react';
import noop from 'lodash/noop';
import Preview from '../Preview';

const code = '<button>OK</button>';

it('should render component renderer', () => {
	const actual = shallow(
		<Preview
			code={code}
			evalInContext={noop}
		/>
	);

	expect(actual).toMatchSnapshot();
});
