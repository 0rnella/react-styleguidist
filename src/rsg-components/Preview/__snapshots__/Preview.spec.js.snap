exports[`test should render component renderer 1`] = `
<div>
  <div />
</div>
`;
